In [56]:
import os
import numpy as np
import pandas as pd
dataset = pd.read_csv('/content/Social_Network_Ads(2).csv')

In [57]:
dataset.head()

User ID  Gender   Age  EstimatedSalary  Purchased
0  15624510    Male  19.0          19000.0          0
1  15810944    Male  35.0          20000.0          0
2  15668575  Female  26.0          43000.0          0
3  15603246  Female  27.0          57000.0          0
4  15804002    Male  19.0          76000.0          0

In [58]:
dataset['Gender'] = dataset['Gender'].astype('category')
dataset['Gender'] = dataset['Gender'].cat.codes
dataset.head()

User ID  Gender   Age  EstimatedSalary  Purchased
0  15624510       1  19.0          19000.0          0
1  15810944       1  35.0          20000.0          0
2  15668575       0  26.0          43000.0          0
3  15603246       0  27.0          57000.0          0
4  15804002       1  19.0          76000.0          0

In [59]:
dataset.drop('User ID' , axis = 1 , inplace = True)

In [60]:
dataset.isnull().sum()

Gender             0
Age                0
EstimatedSalary    0
Purchased          0
dtype: int64

In [61]:
x = dataset.iloc[: , :-1]
y = dataset.iloc[: , -1]

In [62]:
x.head()

Gender   Age  EstimatedSalary
0       1  19.0          19000.0
1       1  35.0          20000.0
2       0  26.0          43000.0
3       0  27.0          57000.0
4       1  19.0          76000.0

In [63]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: Purchased, dtype: int64

In [64]:
from sklearn.model_selection import train_test_split

In [65]:
x_train , x_test , y_train , y_test = train_test_split(x , y , test_size = 0.25 , random_state = 1 , stratify = y)

### Multilayer Perceptron Model (DNN / ANN)

##### Here basically , We build Deep Neural Network / Atrificial Neural Network on tabular or structured data

In [66]:
import tensorflow as tf   # Backend library
from tensorflow import keras # Frontend library
from keras.layers import Dense  # Fully Connected
from keras import Sequential  # Architecture

##### In Multi layer perceptron , 2 architecture / api are there , 1.Functional and 2.Sequential
##### Functional is use when your ask is more than 1 like in same problem statement you have ask about regression , b classification , m classification . in that scenario... more than 1 problem statement
##### Sequential is use when your ask is about 1 like here it ask about purchase or not . It ask about only binary classification . so we will use Sequential architecture

##### We build DNN / MLP Model because we have tabular data

In [67]:
dnn = Sequential()

In [68]:
dnn

<Sequential name=sequential_3, built=False>

#### We build 1st hidden layer

In [69]:
dnn.add(layer = Dense(units = 128 , activation = 'relu' , input_dim = 3))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


#### Here We build 1st hidden layer , where input is 3 (features of dataset) and units(neurons) are 128(we create) and activation fun is 'relu' . always use relu as an activation fun in hidden layers

### 2nd hidden layer

In [70]:
dnn.add(layer = Dense(units = 128 , activation = 'relu'))

#### We create 2nd hidden layer , we create 128 nuerons in 2nd hidden layer and activation is relu because its hidden layer

#### Build ouput layer

In [71]:
dnn.add(layer = Dense(units =1 , activation = 'sigmoid'))

WE create output layer , here we create 1 final neuron , so units is 1 and activation fun is sigmoid . because its output layer means its give output and its binary classification problem so sigmoid . if multiclass classification problem then softmax is activation fun

#### Summary

In [72]:
dnn.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_12 (Dense)                     │ (None, 128)                 │             512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_13 (Dense)                     │ (None, 128)                 │          16,512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_14 (Dense)                     │ (None, 1)                   │             129 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 17,153 (67.00 KB)

 Trainable params: 17,153 (67.00 KB)

 Non-trainable params: 0 (0.00 B)

### Your model is build now just to check How your model get performed

In [73]:
dnn.compile(optimizer = 'sgd' , loss = 'binary_crossentropy' , metrics = ['accuracy'])

##### Here we select optimizer , because optimizer improve performance of model is very fast

#### Now we have to check with performance of model by passing data

In [74]:
dnn.fit(x_train , y_train , validation_data = (x_test , y_test) , epochs = 100)

Epoch 1/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 2s 25ms/step - accuracy: 0.5103 - loss: 13733185536.0000 - val_accuracy: 0.6400 - val_loss: 0.6918
Epoch 2/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6517 - loss: 0.6912 - val_accuracy: 0.6400 - val_loss: 0.6899
Epoch 3/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6332 - loss: 0.6896 - val_accuracy: 0.6400 - val_loss: 0.6880
Epoch 4/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6381 - loss: 0.6877 - val_accuracy: 0.6400 - val_loss: 0.6862
Epoch 5/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6385 - loss: 0.6859 - val_accuracy: 0.6400 - val_loss: 0.6844
Epoch 6/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6383 - loss: 0.6842 - val_accuracy: 0.6400 - val_loss: 0.6827
Epoch 7/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6603 - loss: 0.6807 - val_accuracy: 0.6400 - val_loss: 0.6811
Epoch 8/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.6814 - loss: 0.6768 - val_accuracy:

#### We are not getting good performance because of feature scaling is not did .
#### accuracy: 0.6527 - loss: 0.6470 - val_accuracy: 0.6400 - val_loss: 0.6537
#### Thumb rule , when you build deep neural network feature scaling is necessary

## Now We Build Same Deep Neural Network Model but with Feature Scaling

In [75]:
from sklearn.preprocessing import StandardScaler

In [76]:
scaler = StandardScaler()

In [77]:
x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.transform(x_test)

In [78]:
model = Sequential()

#### Again Build Deep neural network model

In [79]:
model.add(layer = Dense(units = 128 , activation = 'relu' , input_dim = 3))
model.add(layer = Dense(units = 128 , activation = 'relu'))
model.add(layer = Dense(units =1 , activation = 'sigmoid'))
model.summary()

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_15 (Dense)                     │ (None, 128)                 │             512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_16 (Dense)                     │ (None, 128)                 │          16,512 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_17 (Dense)                     │ (None, 1)                   │             129 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 17,153 (67.00 KB)

 Trainable params: 17,153 (67.00 KB)

 Non-trainable params: 0 (0.00 B)

In [80]:
#### Compile the model

In [81]:
model.compile(optimizer = 'sgd' , loss = 'binary_crossentropy' , metrics = ['accuracy'])

In [82]:
#### Fit the model with the data and see performance

In [83]:
model.fit(x_train_scaled , y_train , validation_data = (x_test_scaled , y_test) , epochs = 100)

Epoch 1/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 27ms/step - accuracy: 0.6576 - loss: 0.6833 - val_accuracy: 0.7600 - val_loss: 0.6657
Epoch 2/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7683 - loss: 0.6616 - val_accuracy: 0.8000 - val_loss: 0.6510
Epoch 3/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8109 - loss: 0.6466 - val_accuracy: 0.8200 - val_loss: 0.6369
Epoch 4/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8190 - loss: 0.6271 - val_accuracy: 0.8200 - val_loss: 0.6238
Epoch 5/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8173 - loss: 0.6203 - val_accuracy: 0.8300 - val_loss: 0.6109
Epoch 6/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8048 - loss: 0.6043 - val_accuracy: 0.8200 - val_loss: 0.5986
Epoch 7/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7871 - loss: 0.5948 - val_accuracy: 0.8400 - val_loss: 0.5873
Epoch 8/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8315 - loss: 0.5796 - val_accuracy: 0.8400 - 

#### Now model is showing good perofrmance as compare to previous one . the reason is feature scaling .accuracy: 0.9225 - loss: 0.2465 - val_accuracy: 0.9000 - val_loss: 0.3189